### ISSUE TITLE
White space added above header when using --header-html option

### ISSUE THREAD

**Issue by DaveWishesHe** [wkhtmltopdf/wkhtmltopdf#3974]
**Version:** wkhtmltopdf 0.12.5 (with patched qt)
**Platform:** Ubuntu 16.04.5 (Xenial amd64)

Re-raising this from comments on other closed issues, as the issue is ongoing. It looks like when you add a header using the --header-html option, the header is pushed down using some kind of default spacing. This spacing seems to increase depending on the contents of the header. The same fault applies to the footer when using --footer-html, where the footer is pushed up the page with excess white space.

Here's @JFossey's steps to reproduce https://github.com/wkhtmltopdf/wkhtmltopdf/issues/2042#issuecomment-361898718 

And my own additional notes and demonstration of the spacing increasing depending on the contents of the header html https://github.com/wkhtmltopdf/wkhtmltopdf/issues/2042#issuecomment-399449124

Workaround remains to set the margin-top and margin-bottom values manually.

**Comment by Tomsgu** [comment]
Can you please add an html example(not in an image so we can copy it) and steps to reproduce here? I don't understand why you create a new issue and put details in a comment of closed one, that is different of your issue.

**Comment by DaveWishesHe** [comment]
My apologies, I thought that given that the issue was related to the placement of the header, it was the same bug as #2042. 

The attached archive contains the example html, and also contains the following notes in a README.

[example.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/2213093/example.zip)

There are 3 header files here, header1.html, header2.html and header3.html. Running the following commands will produce 3 pdfs, and in each one the whitespace above the header will increase, seemingly based on the number of new lines in the header file.

header1.html has no new lines in the header, and produces no whitespace above the header.

`wkhtmltopdf --header-html 'http://test.local/header1.html' 'http://test.local/body.html' 'test1.pdf'`

header2.html has some new lines in the header, and produces some whitespace above the header.

`wkhtmltopdf --header-html 'http://test.local/header2.html' 'http://test.local/body.html' 'test2.pdf'`

header3.html has lots of new lines in the header, and produces lots of whitespace above the header.

`wkhtmltopdf --header-html 'http://test.local/header3.html' 'http://test.local/body.html' 'test3.pdf'`

Expected behaviour in all 3 cases would be that there is no whitespace above the header.

**Comment by GregAuger** [comment]
This issue is stopping me from using wkhtmltopdf

**Comment by NoFr1ends** [comment]
Can confirm this error, having this issue too

**Comment by lucas-gp** [comment]
Hello,
I got the same problem : more lines I add to my header, more whitespace is added top of my header...
Someone found a solution to fix it?

**Comment by shanecp** [comment]
I can confirm this issue exists for `footer-html` as well

**Comment by jossef** [comment]
Hey all (and @Tomsgu), I would like to join this party :smile: 

### TL;DR

**I think I've found a way to workaround this issue**. 
1. In your header's html, add `style="position:absolute;"` to your root element `<body>`
1. pass arguments `--header-spacing <guess> --disable-smart-shrinking` 
    - replace `<guess>` with a guessed number (play with it!)


## Workaround

1. In your header's html, add `style="position:absolute;"` to your root element `<body>`

        <!DOCTYPE html>
        <html>
        <head lang="en">
            <meta charset="UTF-8">
            <link href="material-design.css" rel="stylesheet" type="text/css">
            <link href="style.css" rel="stylesheet" type="text/css">
        </head>
        <body>
                <div style="position:absolute;" class="page-header">
                    <img src="header.png">
                </div>
        </body>
        </html>

1. Add to `wkhtmltopdf` command line arguments `--header-spacing <guess> --disable-smart-shrinking` (replace `<guess>` with a guessed number)

       /usr/local/bin/wkhtmltopdf --footer-html /var/.../footer.html --footer-spacing 5 --header-html /var/.../header.html --header-spacing 40 --disable-smart-shrinking --dpi 300 --quiet file://.../content.html /var/.../output.pdf


![problemo](https://user-images.githubusercontent.com/1287098/51915042-d682fb00-23e2-11e9-9015-432ceca0637e.png)


-----

I'm using the latest release (wkhtmltopdf 0.12.5 with patched qt). My OS is ubuntu-server 18.04.01. These are the specific related packages I've installed:
```json
[
    "wkhtmltox_0.12.5-1.bionic_amd64.deb",
    "x11-common_1%3a7.7+19ubuntu7.1_all.deb",
    "xfonts-75dpi_1%3a1.0.4+nmu1_all.deb",
    "libx11-6_2%3a1.6.4-3ubuntu0.1_amd64.deb",
    "fonts-dejavu-core_2.37-1_all.deb",
    "fontconfig_2.12.6-0ubuntu2_amd64.deb",
    "libstdc++6_8.2.0-1ubuntu2~18.04_amd64.deb",
    "libgcc1_1%3a8.2.0-1ubuntu2~18.04_amd64.deb",
    "libjpeg-turbo8_1.5.2-0ubuntu5.18.04.1_amd64.deb",
    "xfonts-encodings_1%3a1.0.4-2_all.deb",
    "zlib1g_1%3a1.2.11.dfsg-0ubuntu2_amd64.deb",
    "fontconfig-config_2.12.6-0ubuntu2_all.deb",
    "libssl1.1_1.1.0g-2ubuntu4.1_amd64.deb",
    "libc6_2.27-3ubuntu1_amd64.deb",
    "ca-certificates_20180409_all.deb",
    "libfontconfig1_2.12.6-0ubuntu2_amd64.deb",
    "xfonts-base_1%3a1.0.4+nmu1_all.deb",
    "libxrender1_1%3a0.9.10-1_amd64.deb",
    "libxext6_2%3a1.3.3-1_amd64.deb",
    "libxcb1_1.13-1_amd64.deb",
    "libfontenc1_1%3a1.1.3-1_amd64.deb",
    "libpng16-16_1.6.34-1ubuntu0.18.04.1_amd64.deb",
    "libfreetype6_2.8.1-2ubuntu2_amd64.deb",
    "gcc-8-base_8.2.0-1ubuntu2~18.04_amd64.deb",
    "xfonts-utils_1%3a7.7+6_amd64.deb"
]
```

Please don't consider this workaround a fix. I would appreciate if this issue will remain opened until the DPI issue is resolved

**Comment by olivierfaure** [comment]
for white space before header and after footer, use --disable-smart-shrinking and no --dpi option, it works well, juste 3mm left as with older versions

**Comment by manojverma086** [comment]
+1

**Comment by mlubimow** [comment]
+1

**Comment by kavishatalsania** [comment]
+1

**Comment by lucas-gp** [comment]
Thanks my friend partulaj to fix it for me :D

**Comment by hg-pooraa** [comment]
Try to change the dpi value. dpi = 80 worked for me. No extra white space above header. You can try your own value by hit and trial. try setting dpi without adding any margin and don't set disable smart shrink.

**Comment by mpth** [comment]
Glad to see this has been fixed. @ashkulz  Any chance this could be released before everything else in 0.12.7 is done?

**Comment by homestar9** [comment]
Has there been any movement on the next release that addresses this issue?

**Comment by Sicos1977** [comment]
https://github.com/Sicos1977/ChromeHtmlToPdf ... also works on Linux
